src/server/pfs/server/driver.go

```diff
@@ -1048,6 +1050,57 @@ func (d *driver) inspectCommit(ctx context.Context, commit *pfs.Commit, block bo
 	return commitInfo, nil
 }
 
+// resolveCommit contains the essential implementation of inspectCommit: it converts 'commit' (which may
+// be a commit ID or branch reference, plus '~' and/or '^') to a repo + commit
+// ID. It accepts an STM so that it can be used in a transaction and avoids an
+// inconsistent call to d.inspectCommit()
+func (d *driver) resolveCommit(stm col.STM, userCommit *pfs.Commit) (resultCommitInfo *pfs.CommitInfo, retErr error) {
+	if userCommit == nil {
+		return nil, fmt.Errorf("cannot resolve nil commit")
+	}
+	commit := userCommit // back up user commit, for error reporting
+	// Extract any ancestor tokens from 'commit.ID' (i.e. ~ and ^)
+	var ancestryLength int
+	commit.ID, ancestryLength = parseCommitID(commit.ID)
+
+	// Check if commit.ID is already a commit ID (i.e. a UUID). Because we use
+	// UUIDv4, the 13th character is a '4'
+	isUUID := len(commit.ID) == uuid.UUIDWithoutDashesLength && commit.ID[12] == '4'
+	if !isUUID {
+		branches := d.branches(commit.Repo.Name).ReadWrite(stm)
+		branchInfo := &pfs.BranchInfo{}
+		// See if we are given a branch
+		if err := branches.Get(commit.ID, branchInfo); err != nil {
+			return nil, err
+		}
+		if branchInfo.Head == nil {
+			return nil, pfsserver.ErrNoHead{branchInfo.Branch}
+		}
+		commit.ID = branchInfo.Head.ID
+	}
+
+	// Traverse commits' parents until you've reached the right ancestor
+	commits := d.commits(commit.Repo.Name).ReadWrite(stm)
+	commitInfo := &pfs.CommitInfo{}
+	for i := 0; i <= ancestryLength; i++ {
+		if commit == nil {
+			return nil, pfsserver.ErrCommitNotFound{userCommit}
+		}
+		childCommit := commit // preserve child for error reporting
+		if err := commits.Get(commit.ID, commitInfo); err != nil {
+			if col.IsErrNotFound(err) {
+				if i == 0 {
+					return nil, pfsserver.ErrCommitNotFound{childCommit}
+				}
+				return nil, pfsserver.ErrParentCommitNotFound{childCommit}
+			}
+			return nil, err
+		}
+		commit = commitInfo.ParentCommit
+	}
+	return commitInfo, nil
+}
```

Comment:
I think we may actually want to expose `resolveCommit` as an rpc in the future, I wrote a similar function in `src/server/pps/server/api_server.go` which just calls `InspectCommit` and only considers the commit it returns. So I think it'd be more efficient and probably make more sense to just use this function. Anyways, nothing necessary for this PR, just figured I'd mention it.

Reply:
makes sense to me :+1: